### PR TITLE
Exclude root and internal parent partitions from age calculation.

### PR DESCRIPTION
**Pull request by ashwinstar**
Root partition and internal parent partitions do not contain any data. But since
currently these tables still have valid relfrozenxid, their age keeps
growing. The problem it poses is to bring the age down need to run vacuum on
root, which trickles down and vacuum's each and every child partition as
well. Instead, if only leaf if getting modified those can be vacuumed in
isolation and bring the age down and avoid the overhead of vacuuming the full
hierarchy.

So, similar to AO, CO, external tables, etc.. for root and parent partitions
record relfrozenxid as 0 (InvalidTransaction) during table creation. This works
since these tables will never store any xids. This would skip them from age
calculation hence eliminate the forced need for vacuum on them. Ideally, same can be
achieved by defining root and internal parent partition as AO tables but its lot
more work and ddls need modifications.

** This PR also contains commit removing guc gp_setwith_alter_storage and corresponding tests. In future when the feature is fully implemented and verified can be exposed and no guc is needed.

### CODE REVIEW

**Comment by ivannovick** [comment]
:) thank you for this

File: src/backend/commands/cluster.c [hunk 1]
```diff
@@ -1402,14 +1396,16 @@ swap_relation_files(Oid r1, Oid r2, bool target_is_pg_class,
 	 * This needs to be performed after the relkind and relstorage has been
 	 * swapped to correctly reflect the relfrozenxid.
 	 */
-	if (should_have_valid_relfrozenxid(relform1->relkind, relform1->relstorage))
+	if (TransactionIdIsValid(relform1->relfrozenxid))
 	{
-		/* set rel1's frozen Xid */
-		Assert(TransactionIdIsNormal(frozenXid));
```
Comment (ashwinstar) [inline src/backend/commands/cluster.c]: will get this assertion back.

**Comment by ashwinstar** [comment]
On Thu, Jul 19, 2018 at 5:17 PM Asim RP <notifications@github.com> wrote:

> Jotting down an alternative approach that just popped up in my mind:
> Always set relfrozenxid to 0 during create. Update it to current xmin at
> the end of first insert, if it's a heap table.
>
As discussed offine, this works only if first insert succeeds. If it aborts
the xid still makes to table and relfroxenxid is out of sync and breaks the
logic.

Need something to track the dirtiness of the table outside of visiblity
rules to help avoid vacuum's on tables which are not being modified but
need to be vacuumed today just for age purpose.

**Comment by ylew-zdata** [comment]
Please ensure that if the root/branch nodes are defined as AO tables that the associated AO metadata tables (visimap, segrelid, etc.) are also accounted for.  Currently in 5.x AO tables do not age but their metadata tables are heap organized and consequently need to be vacuumed.  In addition these metadata tables cannot be vacuumed directly--instead the AO table needs to be vacuumed and the vacuum cascades down.

**Comment by asimrp** [comment]
You have a valid point @ylew-zdata.  I will take care of it in a follow-up patch.  Really, non-leaf partitioned tables deserve a better way to be distinguished from leaf / regular tables.  That way, we can reduce catalog overhead by not creating any of the auxiliary tables / indexes for non-leaf tables.  But that's for a long thoughtful day ...

**Pull request merged**